runtime/memprof.c

```diff
@@ -119,257 +123,416 @@ static uintnat next_mt_generate_binom;
-  value block;
-  value callstack;
-  uintnat occurrences;
-  enum ml_alloc_kind kind;
-} default_postponed_queue[POSTPONED_DEFAULT_QUEUE_SIZE],
-  *postponed_queue = default_postponed_queue,
-  *postponed_queue_end = default_postponed_queue + POSTPONED_DEFAULT_QUEUE_SIZE,
-  *postponed_tl = default_postponed_queue, /* Pointer to next pop */
-  *postponed_hd = default_postponed_queue; /* Pointer to next push */
-
-static struct postponed_block* postponed_next(struct postponed_block* p)
+/* Run all the needed callbacks for a given entry.
+   In case of a thread context switch during a callback, this can be
+   called in a reetrant way.
+   The index of the entry may change. It is set to [Invalid_index] if
+   the entry is discarded.
+   Returns:
+   - An exception result if the callback raised an exception
+   - Val_long(0) == Val_unit == None otherwise
+ */
+static value run_entry_callbacks_exn(uintnat* t_idx)
 {
-  p++;
-  if (p == postponed_queue_end) return postponed_queue;
-  else return p;
-}
+  value sample_info, res, user_data;    /* No need to make these roots */
+  struct tracked* t = &trackst.entries[*t_idx];
+  if (*t_idx == trackst.callback) trackst.callback++;
+
+  if (t->deleted || t->callback_running) return Val_unit;
+
+  if (!t->cb_alloc_called) {
+    t->cb_alloc_called = 1;
+    CAMLassert(Is_block(t->block)
+               || t->block == Placeholder_value
+               || t->deallocated);
+    sample_info = caml_alloc_small(5, 0);
+    Field(sample_info, 0) = Val_long(t->n_samples);
+    Field(sample_info, 1) = Val_long(Wosize_hd(t->header));
+    Field(sample_info, 2) = Val_long(Tag_hd(t->header));
+    Field(sample_info, 3) = Val_long(t->unmarshalled);
+    Field(sample_info, 4) = t->user_data;
+    t->user_data = Val_unit;
+    res = run_callback_exn(t_idx,
+        t->alloc_young ? callback_alloc_minor : callback_alloc_major,
+        sample_info);
+    if (*t_idx == Invalid_index)
+      return res;
+    CAMLassert(!Is_exception_result(res) && Is_block(res) && Tag_val(res) == 0
+               && Wosize_val(res) == 1);
+    t = &trackst.entries[*t_idx];
+    t->user_data = Field(res, 0);
+    if (*t_idx < trackst.young) trackst.young = *t_idx;
+  }
 
-static void purge_postponed_queue(void)
-{
-  if (postponed_queue != default_postponed_queue) {
-    caml_stat_free(postponed_queue);
-    postponed_queue = default_postponed_queue;
-    postponed_queue_end = postponed_queue + POSTPONED_DEFAULT_QUEUE_SIZE;
+  if (t->promoted && !t->cb_promote_called) {
+    t->cb_promote_called = 1;
+    user_data = t->user_data;
+    t->user_data = Val_unit;
+    res = run_callback_exn(t_idx, callback_promote, user_data);
+    if (*t_idx == Invalid_index)
+      return res;
+    CAMLassert(!Is_exception_result(res) && Is_block(res) && Tag_val(res) == 0
+               && Wosize_val(res) == 1);
+    t = &trackst.entries[*t_idx];
+    t->user_data = Field(res, 0);
+    if (*t_idx < trackst.young) trackst.young = *t_idx;
```

Comment:
Would it be useful to check `Is_block(user_data) && Is_young(user_data)` before updating `trackst.young`? (I guess it would make sense for all such `user_data` settings, and it could be encapsulated in a helper function.) My reasoning is that a common case for promotion callbacks will be `fun callstack -> Some callstack`, and the callstack is fairly likely to be already old.

Reply:
Ah, right.